### ISSUE TITLE
Can events carry some custom parameters?

### ISSUE THREAD

**Issue by smile-coder** [derkork/godot-statecharts#135]
Hello, the state diagram is really easy to use and great tool. I found that the triggered event can not carry some custom parameters. I would like to know if this can be implemented. If it can be implemented, it will greatly increase the ease of use. For example, if a player takes damage in a non-invincible state, the damage value can be customized by triggering the function: func event_received(event, **args**...).

**Comment by derkork** [comment]
Funny enough I had a similar idea the other day. This is most certainly doable though it will be a breaking change because of the signal. It will also only support a single argument but that should be sufficient as you can give any Godot object as event data, which includes custom objects and things like dictionaries and arrays. It will probably be a bit more work though as I think it would be very useful to have event data also available in expression guards and this will need a bit of fiddling.

**Comment by eobet** [comment]
I must admit I'm confused and hesitant about any sort of code based logic in Godot State Charts, and that included the Expression Guard.

However, I too would love to have something like this, because I'm currently having quite some trouble figuring out how to solve the particular type of movement logic I want without a more explicit queue or some sort or a signal/event guard...

If this feature would be expanded upon, I'd love it if custom parameters could also be managed in a list, just like events now are!

**Comment by DaveMS** [comment]
I would also like to see this, and jumped on to here to post the same thing. 

in my case I want to transition to a "Switching Weapon" state when a user presses a key to switch to a different weapon. 
It would be useful to be able to pass a parameter to the new state to specify which weapon is being switched to so that the new state can load the correct model and perform the appropriate animation.

At the moment I have to work around it by storing this in a class level `weapon_to_switch_to` variable which is only used to pass this data between these 2 states.

**Comment by DaveMS** [comment]
Thinking about it a bit more, maybe my usecase is slightly different than the one posted. 
But as an extension to the initial enhancement and custom parameters are added to events: if a state transition is triggered by an event with a custom parameter, would it be possible to pass the custom parameter data to the `on_entered` handler of the new state?

**Comment by derkork** [comment]
Yes that would indeed make sense, however this is going to break every project that is currently using the library as I cannot change the signal parameter list in a backwards compatible way. I will have to think about how I can make this happen in a way that isn't super frustrating when upgrading the project to a new library version. Also there are some interesting edge cases with this setup. Say you have a setup like this:

![image](https://github.com/user-attachments/assets/3389d3c0-7fe2-48b1-a3a5-6cdc26dc7823)

Now you do a `send_event("some_event", some_value)`. This will call `state_entered` on _B_ and give you `some_value` as a parameter. Then immediately after this _B_ will transition to _C_ or _D_ depending on the value of `x`. Now say it transitions to _C_, what is given as a parameter to `state_entered` for _C_? Is is `null` because the state change wasn't technically directly triggered by an event or is it `some_value` because the state change was a chain effect from the transition to _B_. I think you can find valid arguments for both positions but I really can only implement one of them (also detecting chain effects may be tricky). So I think this is going to need a bit more thought to have a universally useful behaviour.

**Comment by Zehir** [comment]
Just found this rabbit hole like everybody, I also want to have some parameters inside the event or some kind of state context like in [xstate](https://stately.ai/docs/context).
But for this addon there is no context except the `state_chart.set_expression_property` used for expressions guards on transitions.
So my first idea was instead of sending event I could update the data using `set_expression_property` function and having an automatic transition that react to it.

For example I have a state `Idle`. I set the value `target_tile` and the transition see that `target_tile` is not null anymore so the new state is `Moving`. Then I could send the `Stoped` event to return back to the `Idle` state. But before sending that event I need to set the `target_tile` to null or I will get an infinite loop.

An other idea was to have a global object that I update juste before sending event. But if my event is triggered later something else could have updated the context before the event was processed.
That probably why xstate update his context is an action that is processed during the event processing.

An other idea is also to have border script where you don't directly send events to the statechart but you have a man in the middle script. That script store the context and send event to the child statechart. And he have his own event queue to delay new event for entering the state machine until her event queue is empty.
